arm-software/shared/elf2bin/elf2bin.cpp

```diff
@@ -0,0 +1,468 @@
+    inobj.elf = std::move(elf);
+  }
+
+  /*
+   * Helper function for listing the segments of a file, paying
+   * attention to the --segments option to restrict to a subset.
+   */
+  auto segments = [&](InputObject &inobj) {
+    std::vector<Segment> allsegs = inobj.segments(physical);
+    if (!segments_wanted)
+      return allsegs;
+
+    std::vector<Segment> segs;
+    auto &keep = segments_wanted.value();
+    for (auto seg : allsegs)
+      if (keep.find(seg.baseaddr) != keep.end())
+        segs.push_back(seg);
+    return segs;
+  };
+
+  /*
+   * Make a list of all the conversions we want to do.
+   */
+
+  struct Conversion {
+    InputObject *inobj;
+    std::string outfile;
+    std::optional<uint64_t> baseaddr, fileoffset, size, zisize;
+    uint64_t bank;
+  };
+  std::vector<Conversion> convs;
+
+  for (auto &inobj : objects) {
+    // Helper function to fill in infile and outfile
+    auto add_conv = [&](Conversion conv) {
+      conv.inobj = &inobj;
+      if (outfile)
+        conv.outfile = outfile.value();
+      else
+        conv.outfile = format_outfile(outpattern.value(), conv.inobj->filename,
+                                      conv.baseaddr, conv.bank);
+      convs.push_back(conv);
+    };
+
+    switch (format.value()) {
+    case Format::BinMultifile:
+    case Format::VhxMultifile:
+      /*
+       * Separate output file per segment and per bank, so go
+       * through this input file and list its segments.
+       */
+      for (auto seg : segments(inobj)) {
+        for (uint64_t bank = 0; bank < nbanks; bank++) {
+          Conversion conv;
+          conv.baseaddr = seg.baseaddr;
+          conv.fileoffset = seg.fileoffset;
+          conv.size = seg.filesize;
+          conv.bank = bank;
+          if (include_zi && seg.memsize > seg.filesize)
+            conv.zisize = seg.memsize - seg.filesize;
+          else
+            conv.zisize = 0;
+          add_conv(conv);
+        }
+      }
+      break;
+    case Format::BinCombined:
+    case Format::VhxCombined:
+      /*
+       * Separate output file per bank, but each one contains
+       * the whole input file.
+       */
+      for (uint64_t bank = 0; bank < nbanks; bank++) {
+        Conversion conv;
+        conv.bank = bank;
+        add_conv(conv);
+      }
+      break;
+    default:
+      /*
+       * One output file per input file.
+       */
+      add_conv(Conversion{});
+      break;
+    }
+  }
+
+  std::set<std::string> outfiles;
+  for (const auto &conv : convs) {
+    if (outfiles.find(conv.outfile) != outfiles.end()) {
+      fatal(Twine("output file '") + conv.outfile +
+            "' would be written more than once by this command");
+      std::exit(1);
+    }
+    outfiles.insert(conv.outfile);
+  }
+
+  uint64_t bankmod = nbanks * bankwidth;
+
+  for (const auto &conv : convs) {
+    switch (format.value()) {
+    case Format::BinMultifile:
+      bin_write(*conv.inobj, conv.outfile, conv.fileoffset.value(),
```

Comment:
The functions have a lot of shared parameters, could it be worth passing conv in as a "context" parameter.

Alternatively we should be doing just one of the options in this switch for each of the convs. Could we extend conv to be a transformation object that includes a write() member function?

Reply:
I should rephrase to be less specific to conversion. There are quite a few parameters that are shared between all of the functions and these could be grouped together in a struct to shorten the parameter list. Given that some of these functions have upwards of 8 parameters it can be hard to spot differences.

Overall this is just a subjective preference on my part.